src/condor_utils/proc_family_direct_cgroup_v1.cpp

```diff
@@ -0,0 +1,485 @@
+// putting the job into the blkio controller will give
+// job-level fair share to filesystem io.
+static std::array<const std::string, 4> controllers {
+		"memory", "cpu,cpuacct", "freezer", "blkio"};
+
+// Note that even root can't remove control group *files*,
+// only directories.
+void fullyRemoveCgroup(const stdfs::path &absCgroup) {
+
+	// Common case, it shouldn't still exist, we're done
+	if (!stdfs::exists(absCgroup)) {
+		return;
+	}
+
+	// Otherwise, we need to depth-firstly remove any subdirs
+	for (auto const& entry: std::filesystem::directory_iterator{absCgroup})
+    {
+		if (entry.is_directory()) {
+			// Must go depth-first
+			fullyRemoveCgroup(absCgroup / entry);
+			int r = rmdir((absCgroup / entry).c_str());
+			if ((r < 0) && (errno != ENOENT))  {
+				dprintf(D_ALWAYS, "ProcFamilyDirectCgroupV1 error removing cgroup %s: %s\n", (absCgroup / entry).c_str(), strerror(errno));
+			} else {
+				dprintf(D_FULLDEBUG, "ProcFamilyDirect removed old cgroup %s\n", (absCgroup / entry).c_str());
+			}
+		}
+	}
+	// Finally, remove "."
+	int r = rmdir(absCgroup.c_str());
+	if ((r < 0) && (errno != ENOENT))  {
+		dprintf(D_ALWAYS, "ProcFamilyDirectCgroupV1 error removing cgroup %s: %s\n", absCgroup.c_str(), strerror(errno));
+	} else {
+		dprintf(D_FULLDEBUG, "ProcFamilyDirect removed old cgroup %s\n", absCgroup.c_str());
+	}
+}
+
+// mkdir the cgroup, and all required interior cgroups.
+bool 
+ProcFamilyDirectCgroupV1::cgroupify_process(const std::string &cgroup_name, pid_t pid) {
+	dprintf(D_FULLDEBUG, "Creating cgroup %s for pid %d\n", cgroup_name.c_str(), pid);
+
+	TemporaryPrivSentry sentry( PRIV_ROOT );
+
+	// Start from the root of the cgroup mount point
+	stdfs::path cgroup_root_dir = cgroup_mount_point();
+
+	for (const std::string &controller: controllers) {
+		stdfs::path absolute_cgroup = cgroup_root_dir / stdfs::path(controller) / cgroup_name;
+
+		// If the full cgroup exists, remove it to clear the various
+		// peak statistics and any existing memory
+		// GGT TODO should be remove_fully (also in v2)
+		fullyRemoveCgroup(absolute_cgroup);
+
+		bool can_make_cgroup_dir = mkdir_and_parents_if_needed(absolute_cgroup.c_str(), 0755, 0755, PRIV_ROOT);
+		if (!can_make_cgroup_dir) {
+			dprintf(D_ALWAYS, "Cannot mkdir %s, failing to use cgroups\n", absolute_cgroup.c_str());
+			return false;
+		}
+
+		// Now move pid to the leaf of the newly-created tree
+		stdfs::path procs_filename = absolute_cgroup / "cgroup.procs";
+		int fd = open(procs_filename.c_str(), O_WRONLY, 0666);
+		if (fd > 0) {
+			std::string buf;
+			formatstr(buf, "%u", pid);
+			int r = write(fd, buf.c_str(), strlen(buf.c_str()));
+			if (r < 0) {
+				dprintf(D_ALWAYS, "Error writing procid %d to %s: %s\n", pid, procs_filename.c_str(), strerror(errno));
+				close(fd);
+				return false;
+			} else {
+				dprintf(D_ALWAYS, "Moved process %d to cgroup %s\n", pid, absolute_cgroup.c_str());
+			}
+			close(fd);
+		}
```

Comment:
So if we fail to open `cgroup.procs` and therefore never actually assign the process to the cgroup, it's OK?  What am I missing, here?

Reply:
Ah, good catch.  Change pushed.